pkg/engine/lifecycletest/pending_replace_test.go

```diff
@@ -0,0 +1,487 @@
+
+	upHostF := deploytest.NewPluginHostF(nil, nil, programF, upLoaders...)
+	upOptions := TestUpdateOptions{T: t, HostF: upHostF}
+
+	upSnap, err := TestOp(Update).
+		RunStep(project, p.GetTarget(t, nil), upOptions, false, p.BackendClient, nil, "0")
+	assert.NoError(t, err)
+
+	assert.Len(t, upSnap.Resources, 2)
+	assert.Equal(t, upSnap.Resources[0].URN.Name(), "default")
+	assert.Equal(t, upSnap.Resources[1].URN.Name(), "resA")
+
+	// Operation 2 -- return a replacing diff and interrupt it with a failing
+	// create.
+	replaceLoaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{
+				DiffF:   returnReplaceDiff,
+				DeleteF: trackedDelete,
+				CreateF: throwingCreate,
+			}, nil
+		}),
+	}
+
+	replaceHostF := deploytest.NewPluginHostF(nil, nil, programF, replaceLoaders...)
+	replaceOptions := TestUpdateOptions{T: t, HostF: replaceHostF}
+
+	replaceSnap, err := TestOp(Update).
+		RunStep(project, p.GetTarget(t, upSnap), replaceOptions, false, p.BackendClient, nil, "1")
+	assert.ErrorContains(t, err, "interrupt replace")
+
+	assert.Len(t, replaceSnap.Resources, 2)
+	assert.Equal(t, replaceSnap.Resources[0].URN.Name(), "default")
+	assert.Equal(t, replaceSnap.Resources[1].URN.Name(), "resA")
+	assert.True(t, deleteCalled, "Delete should be called as part of replacement")
+	assert.True(t, replaceSnap.Resources[1].PendingReplacement)
+
+	// Operation 3 -- resume the replacement with the same program.
+	deleteCalled = false
+	createCalled = false
+
+	removeLoaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{
+				DiffF:   returnReplaceDiff,
+				DeleteF: trackedDelete,
+				CreateF: trackedCreate,
+			}, nil
+		}),
+	}
+
+	removeHostF := deploytest.NewPluginHostF(nil, nil, programF, removeLoaders...)
+	removeOptions := TestUpdateOptions{T: t, HostF: removeHostF}
+
+	removeSnap, err := TestOp(Update).
+		RunStep(project, p.GetTarget(t, replaceSnap), removeOptions, false, p.BackendClient, nil, "2")
+	assert.NoError(t, err)
+
+	// Assert.
+	assert.Len(t, removeSnap.Resources, 2)
+	assert.Equal(t, removeSnap.Resources[0].URN.Name(), "default")
+	assert.Equal(t, removeSnap.Resources[1].URN.Name(), "resA")
+	assert.False(t, deleteCalled, "Delete shouldn't be called a second time when resuming a replacement (same goals)")
+	assert.True(t, createCalled, "Create should be called when resuming a replacement (same goals)")
+	assert.False(t, removeSnap.Resources[1].PendingReplacement)
+}
+
+// Tests that a delete-before-replace operation:
+//
+// * that is interrupted after the deletion (e.g. with a failed create)
+// * and then resumed, with a new program that removes the deleted resource
+//
+// will:
+//
+// * not call delete (again) on the old resource
+// * remove the PendingReplacement flag from the resource in state
```

Comment:
Isn't this the same thing? Like if the resource is removed from state does it matter what we do with the flag?

Reply:
Yea true. This is a detail of how the steps look. Will remove this line.